core/schema/src/main/liquibase/21.0.0/changelog.xml

```diff
@@ -111,5 +111,46 @@
     <dropColumn tableName="snmpinterface" columnName="snmpipadentnetmask" />
   </changeSet>
 
+  <!-- See NMS-9742 -->
+  <changeSet id="21.0.2-remove-duplicate-reductionkey-memos" author="mvrueden">
+    <!-- Merge existing duplicate entries -->
+    <sql>
+      WITH subquery AS (
+      SELECT
+      reductionkey,
+      type,
+      max(updated) AS updated,
+      string_agg(body, E'\n\nMerged by database installer - ' || to_char(current_timestamp, 'YYYY-MM-DD') || E':\n') AS body,
+      count(*)
+      FROM memos
+      WHERE reductionkey IS NOT NULL
+      GROUP BY reductionkey, type
+      HAVING count(*) > 1
+      )
+      UPDATE memos
+      SET body = subquery.body
+      FROM subquery
+      WHERE memos.reductionkey = subquery.reductionkey
+      AND memos.type = subquery.type
+      AND memos.updated = subquery.updated;
+    </sql>
+    <!-- Delete duplicate entries except the newest -->
+    <sql>
+      DELETE FROM memos WHERE id IN (
```

Comment:
I think there's a problem with the delete here, if I run:
```
delete from memos;
insert into memos (id,created,updated,author,body,reductionkey,type) VALUES (1,NOW(),NOW(),'me','body1','rkey1','note');
insert into memos (id,created,updated,author,body,reductionkey,type) VALUES (2,NOW(),NOW(),'me','body2','rkey1','note');
insert into memos (id,created,updated,author,body,reductionkey,type) VALUES (3,NOW(),NOW(),'me','body3','rkey2','note');
insert into memos (id,created,updated,author,body,reductionkey,type) VALUES (4,NOW(),NOW(),'me','body4','rkey2','note');
insert into memos (id,created,updated,author,body,reductionkey,type) VALUES (5,NOW(),NOW(),'me','body5','rkey3','note');
```

and then execute these two SQL statements, I'm left with 2 entries in the memos table, whereas I would expect there to be 3.

Reply:
Yes, you are right. The SQL only works if you have only one duplicate. Good thing you caught that. It could have created a huge mess without anyone noticing. I changed the SQL to work properly.